File: Depth-First Search/Binary tree/Find paths/Easy/LET_no257_Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        if (!root)
            return res;

        string s = to_string(root->val);
        dfs(root, s, res);
        return res;
    }

    void dfs(TreeNode* root, string& s, vector<string>& res) {
        if (!root->left && !root->right) {
            res.push_back(s);
            return;
        }

        if (root->left) {
            s += "->" + to_string(root->left->val);
            dfs(root->left, s, res);
            size_t pos = s.rfind("->");
            s.erase(pos, s.size() - pos);
        }

        if (root->right) {
            s += "->" + to_string(root->right->val);
            dfs(root->right, s, res);
            size_t pos = s.rfind("->");
            s.erase(pos, s.size() - pos);
        }
    }
};

File: Depth-First Search/Binary tree/Find paths/Easy/LET_no257_Binary Tree Paths.py
# The same as LeetCode no257. Binary Tree Paths

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of the binary tree
    @return: all root-to-leaf paths
    """
    def binaryTreePaths(self, root):

        if not root:
            return []

        result = []
        self.dfs(root, [str(root.val)], result)
        return result

    def dfs(self, root, path, result):

        if not root.left and not root.right:
            result.append("->".join(path))
            return

        if root.left:
            path.append(str(root.left.val))
            self.dfs(root.left, path, result)
            path.pop()

        if root.right:
            path.append(str(root.right.val))
            self.dfs(root.right, path, result)
            path.pop()

        return